search-books/src/main/java/nix/education/service/TransactionalDataService.java

```diff
@@ -0,0 +1,38 @@
+package nix.education.service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import nix.education.entity.Book;
+import nix.education.entity.Tag;
+import nix.education.repository.BookRepository;
+import nix.education.repository.TagRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TransactionalDataService {
+
+    public List<String> getTagTitles(Book book) {
+        return book.getTags().stream().map(Tag::getTag).collect(Collectors.toList());
+    }
+
+    public List<Book> getBooksByIds(List<Long> ids, BookRepository bookRepository) {
+        List<Book> books = new ArrayList<>();
+        ids.forEach(id -> books.add(bookRepository.findById(id).orElseThrow()));
+        return books;
+    }
+
+    public List<Tag> getResolvedTags(Book book, TagRepository tagRepository) {
+        List<Tag> tags = new ArrayList<>();
+
+        for (Tag tag : book.getTags()) {
+            if (!tagRepository.existsTagByTag(tag.getTag())) {
```

Comment:
For each tag value we make a call to DB. Not good from the performance standpoint

Reply:
Changed logic for searching existed tags.